core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java

```diff
@@ -0,0 +1,424 @@
+    if (master == null) {
+      throw new NotFoundException(ExceptionMessage.NO_MASTER_FOUND.getMessage(masterId));
+    }
+
+    master.updateLastUpdatedTimeMs();
+
+    List<ConfigProperty> configList = options.getConfigList().stream()
+        .map(ConfigProperty::fromThrift).collect(Collectors.toList());
+    mMasterConfigStore.registerNewConf(master.getAddress(), configList);
+
+    LOG.info("registerMaster(): master: {} options: {}", master, options);
+  }
+
+  /**
+   * Lost master periodic check.
+   */
+  private final class LostMasterDetectionHeartbeatExecutor implements HeartbeatExecutor {
+
+    /**
+     * Constructs a new {@link LostMasterDetectionHeartbeatExecutor}.
+     */
+    public LostMasterDetectionHeartbeatExecutor() {
+    }
+
+    @Override
+    public void heartbeat() {
+      int masterTimeoutMs = (int) Configuration.getMs(PropertyKey.MASTER_HEARTBEAT_TIMEOUT_MS);
+      for (MasterInfo master : mMasters) {
+        synchronized (master) {
+          final long lastUpdate = mClock.millis() - master.getLastUpdatedTimeMs();
+          if (lastUpdate > masterTimeoutMs) {
+            LOG.error("The master {}({}) timed out after {}ms without a heartbeat!", master.getId(),
+                master.getAddress(), lastUpdate);
+            mLostMasters.add(master);
+            mMasters.remove(master);
+            mMasterConfigStore.handleNodeLost(master.getAddress());
+          }
+        }
+      }
+    }
+
+    @Override
+    public void close() {
+      // Nothing to clean up
+    }
+  }
+
+  /**
+   * Updates the config checker when a lost worker becomes alive.
+   *
+   * @param address the address of the worker
+   */
+  private void lostWorkerFoundHandler(Address address) {
+    mWorkerConfigStore.lostNodeFound(address);
+  }
+
+  /**
+   * Updates the config checker when a live worker becomes lost.
+   *
+   * @param address the address of the worker
+   */
+  private void workerLostHandler(Address address) {
+    mWorkerConfigStore.handleNodeLost(address);
+  }
+
+  /**
+   * Updates the config checker when a worker registers with configuration.
+   *
+   * @param address the address of the worker
+   * @param configList the configuration of this worker
+   */
+  private void registerNewWorkerConfHandler(Address address, List<ConfigProperty> configList) {
+    mWorkerConfigStore.registerNewConf(address, configList);
+  }
+
+  /**
+   * Sets the master id. This method should only be called when this master is a standby master.
+   *
+   * @param metaMasterClient the meta master client to communicate with leader master
+   */
+  private void setMasterId(MetaMasterMasterClient metaMasterClient) {
+    try {
+      RetryUtils.retry("get master id",
+          () -> mMasterId.set(metaMasterClient.getId(mMasterAddress)),
+          ExponentialTimeBoundedRetry.builder()
+              .withMaxDuration(Duration
+                  .ofMillis(Configuration.getMs(PropertyKey.USER_RPC_RETRY_MAX_NUM_RETRY)))
```

Comment:
this should probably wait indefinitely instead of crashing the master. Perhaps add a new property key with a super-long default? @